File: src/connection/core/options.rs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use std::time::Duration;

use typedb_protocol::{
    options::{
        ExplainOpt::Explain, InferOpt::Infer, ParallelOpt::Parallel, PrefetchOpt::Prefetch,
        PrefetchSizeOpt::PrefetchSize, ReadAnyReplicaOpt::ReadAnyReplica,
        SchemaLockAcquireTimeoutOpt::SchemaLockAcquireTimeoutMillis,
        SessionIdleTimeoutOpt::SessionIdleTimeoutMillis, TraceInferenceOpt::TraceInference,
        TransactionTimeoutOpt::TransactionTimeoutMillis,
    },
    Options as OptionsProto,
};

macro_rules! options {
    {pub struct $name:ident { $(pub $field_name:ident : Option<$field_type:ty>),* $(,)? }} => {
        #[derive(Clone, Debug, Default)]
        pub struct $name {
            $(pub $field_name: Option<$field_type>,)*
        }

        impl $name {
            $(
            pub fn $field_name(mut self, value: $field_type) -> Self {
                self.$field_name = value.into();
                self
            }
            )*
        }
    };
}

options! {
    pub struct Options {
        pub infer: Option<bool>,
        pub trace_inference: Option<bool>,
        pub explain: Option<bool>,
        pub parallel: Option<bool>,
        pub prefetch: Option<bool>,
        pub prefetch_size: Option<i32>,
        pub session_idle_timeout: Option<Duration>,
        pub transaction_timeout: Option<Duration>,
        pub schema_lock_acquire_timeout: Option<Duration>,
    }
}

options! {
    pub struct ClusterOptions {
        pub infer: Option<bool>,
        pub trace_inference: Option<bool>,
        pub explain: Option<bool>,
        pub parallel: Option<bool>,
        pub prefetch: Option<bool>,
        pub prefetch_size: Option<i32>,
        pub session_idle_timeout: Option<Duration>,
        pub transaction_timeout: Option<Duration>,
        pub schema_lock_acquire_timeout: Option<Duration>,
        pub read_any_replica: Option<bool>,
    }
}

impl Options {
    pub fn new_core() -> Options {
        Options::default()
    }

    pub fn new_cluster() -> ClusterOptions {
        ClusterOptions::default()
    }

    pub(crate) fn to_proto(&self) -> OptionsProto {
        OptionsProto {
            infer_opt: self.infer.map(Infer),
            trace_inference_opt: self.trace_inference.map(TraceInference),
            explain_opt: self.explain.map(Explain),
            parallel_opt: self.parallel.map(Parallel),
            prefetch_size_opt: self.prefetch_size.map(PrefetchSize),
            prefetch_opt: self.prefetch.map(Prefetch),
            session_idle_timeout_opt: self
                .session_idle_timeout
                .map(|val| SessionIdleTimeoutMillis(val.as_millis() as i32)),
            transaction_timeout_opt: self
                .transaction_timeout
                .map(|val| TransactionTimeoutMillis(val.as_millis() as i32)),
            schema_lock_acquire_timeout_opt: self
                .schema_lock_acquire_timeout
                .map(|val| SchemaLockAcquireTimeoutMillis(val.as_millis() as i32)),
            read_any_replica_opt: None,
        }
    }
}

impl ClusterOptions {
    pub(crate) fn to_proto(&self) -> OptionsProto {
        OptionsProto {
            infer_opt: self.infer.map(Infer),
            trace_inference_opt: self.trace_inference.map(TraceInference),
            explain_opt: self.explain.map(Explain),
            parallel_opt: self.parallel.map(Parallel),
            prefetch_size_opt: self.prefetch_size.map(PrefetchSize),
            prefetch_opt: self.prefetch.map(Prefetch),
            session_idle_timeout_opt: self
                .session_idle_timeout
                .map(|val| SessionIdleTimeoutMillis(val.as_millis() as i32)),
            transaction_timeout_opt: self
                .transaction_timeout
                .map(|val| TransactionTimeoutMillis(val.as_millis() as i32)),
            schema_lock_acquire_timeout_opt: self
                .schema_lock_acquire_timeout
                .map(|val| SchemaLockAcquireTimeoutMillis(val.as_millis() as i32)),
            read_any_replica_opt: self.read_any_replica.map(ReadAnyReplica),
        }
    }
}

File: cpp/lib/connection/options.cpp
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <optional>

#include "typedb/common/exception.hpp"
#include "typedb/connection/options.hpp"

#include "inc/macros.hpp"

namespace TypeDB {

Options::Options()
    : optionsNative(_native::options_new(), _native::options_drop) {}

_native::Options* Options::getNative() const {
    return optionsNative.get();
}

std::optional<bool> Options::infer() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_infer(optionsNative.get()) ? std::optional<bool>(_native::options_get_infer(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<bool> Options::traceInference() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_trace_inference(optionsNative.get()) ? std::optional<bool>(_native::options_get_trace_inference(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<bool> Options::explain() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_explain(optionsNative.get()) ? std::optional<bool>(_native::options_get_explain(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<bool> Options::parallel() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_parallel(optionsNative.get()) ? std::optional<bool>(_native::options_get_parallel(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<bool> Options::prefetch() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_prefetch(optionsNative.get()) ? std::optional<bool>(_native::options_get_prefetch(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<int32_t> Options::prefetchSize() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_prefetch_size(optionsNative.get()) ? std::optional<int32_t>(_native::options_get_prefetch_size(optionsNative.get())) : std::optional<int32_t>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<int64_t> Options::sessionIdleTimeoutMillis() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_session_idle_timeout_millis(optionsNative.get()) ? std::optional<int64_t>(_native::options_get_session_idle_timeout_millis(optionsNative.get())) : std::optional<int64_t>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<int64_t> Options::transactionTimeoutMillis() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_transaction_timeout_millis(optionsNative.get()) ? std::optional<int64_t>(_native::options_get_transaction_timeout_millis(optionsNative.get())) : std::optional<int64_t>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<int64_t> Options::schemaLockAcquireTimeoutMillis() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_schema_lock_acquire_timeout_millis(optionsNative.get()) ? std::optional<int64_t>(_native::options_get_schema_lock_acquire_timeout_millis(optionsNative.get())) : std::optional<int64_t>();
    TypeDBDriverException::check_and_throw();
    return t;
}

std::optional<bool> Options::readAnyReplica() {
    CHECK_NATIVE(optionsNative);
    auto t = _native::options_has_read_any_replica(optionsNative.get()) ? std::optional<bool>(_native::options_get_read_any_replica(optionsNative.get())) : std::optional<bool>();
    TypeDBDriverException::check_and_throw();
    return t;
}

Options& Options::infer(bool infer) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_infer(optionsNative.get(), infer);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::traceInference(bool traceInference) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_trace_inference(optionsNative.get(), traceInference);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::explain(bool explain) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_explain(optionsNative.get(), explain);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::parallel(bool parallel) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_parallel(optionsNative.get(), parallel);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::prefetch(bool prefetch) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_prefetch(optionsNative.get(), prefetch);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::prefetchSize(int32_t prefetchSize) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_prefetch_size(optionsNative.get(), prefetchSize);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::sessionIdleTimeoutMillis(int64_t sessionIdleTimeoutMillis) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_session_idle_timeout_millis(optionsNative.get(), sessionIdleTimeoutMillis);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::transactionTimeoutMillis(int64_t transactionTimeoutMillis) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_transaction_timeout_millis(optionsNative.get(), transactionTimeoutMillis);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::schemaLockAcquireTimeoutMillis(int64_t schemaLockAcquireTimeoutMillis) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_schema_lock_acquire_timeout_millis(optionsNative.get(), schemaLockAcquireTimeoutMillis);
    TypeDBDriverException::check_and_throw();
    return *this;
}

Options& Options::readAnyReplica(bool readAnyReplica) {
    CHECK_NATIVE(optionsNative);
    _native::options_set_read_any_replica(optionsNative.get(), readAnyReplica);
    TypeDBDriverException::check_and_throw();
    return *this;
}
}  // namespace TypeDB
